lib/util/LocationStatusManager.js

```diff
@@ -59,11 +203,476 @@ class LocationStatusManager {
+                stateMachine.removePendingOperation(operation.id);
+                return;
+            }
+            await this._executeOperation(operation, stateMachine);
+            this._logger.debug('Operation executed successfully', {
+                method: 'LocationStatusManager._processOperation',
+                operationId: operation.id,
+                type: operation.type,
+                service: operation.service,
+                location: operation.location,
+            });
+            operation.setState(ServiceLocationOperation.STATES.COMPLETED);
+            stateMachine.removePendingOperation(operation.id);
+            
+            this._logger.debug('Operation completed successfully', {
+                method: 'LocationStatusManager._processOperation',
+                operationId: operation.id,
+                type: operation.type,
+                service: operation.service,
+                location: operation.location,
+            });
+
+    } catch (error) {
+            operation.setState(ServiceLocationOperation.STATES.FAILED, error);
+            this._logger.error('Operation failed', {
+                method: 'LocationStatusManager._processOperation',
+                operationId: operation.id,
+                error: error.message,
+                retryCount: operation.retryCount,
+            });
+
+            if (operation.retryCount < operation.maxRetries) {
+                setTimeout(() => {
+                    this._operationQueue.push(operation);
+                }, 1000 * Math.pow(2, operation.retryCount));
+            } else {
+                operation.setState(ServiceLocationOperation.STATES.CANCELLED, error);
+                stateMachine.removePendingOperation(operation.id);
+            }
+        }
+    }
+
+    /**
+     * Wait for operations to complete with timeout
+     * @param {Array} operations - Operations to wait for
+     * @param {number} timeoutMs - Timeout in milliseconds
+     */
+    async _waitForOperations(operations, timeoutMs = OPERATION_WAIT_TIMEOUT_MS) {
+        const startTime = Date.now();
+        
+        while (operations.some(op => op.state === ServiceLocationOperation.STATES.PROCESSING)) {
+            if (Date.now() - startTime > timeoutMs) {
+                throw new Error('Timeout waiting for operations to complete');
+            }
+            await delay(COMPLETION_POLL_INTERVAL_MS);
+        }
+    }
+
+        /**
+     * Execute the actual operation
+     */
+    async _executeOperation(operation, stateMachine) {
+        const { type, service, location, options } = operation;
+    
+        if (this._serviceConfig[service] && !this._serviceConfig[service].isMongo) {
+            this._pushActionToRedis(service, location, 
+                LocationStatusManager.getRedisAction(type), options.scheduleDate);
+            stateMachine.executeStateTransition(type, options);
+            return;
+        }
+        switch (type) {
```

Comment:
one of the purpose of FSM is to avoid these kind of switch: if there is some operation to performed, it should be part of the FSM definition (linked to a transition or entry into the state, but there anyway)